File: implementations/go/aep/bridge/mcp_bridge_test.go
package bridge

import (
	"sync"
	"testing"
	"time"

	"github.com/axisrobo/aep/aep"
)

func TestMcpInitialize(t *testing.T) {
	bridge := NewMcpBridge(nil)
	resp := bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": map[string]any{}})
	result := resp["result"].(map[string]any)
	info := result["serverInfo"].(map[string]any)
	if info["name"] != "aep-mcp-bridge" {
		t.Fatalf("expected aep-mcp-bridge, got %v", info["name"])
	}
	if result["protocolVersion"] != "0.1.0" {
		t.Fatalf("expected 0.1.0, got %v", result["protocolVersion"])
	}
}

func TestMcpNotificationsInitialized(t *testing.T) {
	bridge := NewMcpBridge(nil)
	resp := bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "method": "notifications/initialized"})
	if resp != nil {
		t.Fatalf("expected nil, got %v", resp)
	}
}

func TestMcpToolsList(t *testing.T) {
	bridge := NewMcpBridge(nil)
	bridge.RegisterTool(McpTool{
		Name:   "echo",
		Schema: map[string]any{"description": "echo tool", "required": []any{"text"}},
		Handler: func(args map[string]any, ctx McpContext) map[string]any {
			return map[string]any{"content": []any{map[string]any{"type": "text", "text": args["text"]}}}
		},
	})
	resp := bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "id": 2, "method": "tools/list"})
	result := resp["result"].(map[string]any)
	tools := result["tools"].([]map[string]any)
	if len(tools) != 1 || tools[0]["name"] != "echo" {
		t.Fatalf("expected echo tool, got %v", tools)
	}
}

func TestMcpToolsCall(t *testing.T) {
	bridge := NewMcpBridge(nil)
	bridge.RegisterTool(McpTool{
		Name:   "echo",
		Schema: map[string]any{"description": "echo"},
		Handler: func(args map[string]any, ctx McpContext) map[string]any {
			return map[string]any{"content": []any{map[string]any{"type": "text", "text": args["text"]}}}
		},
	})
	resp := bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "id": 3, "method": "tools/call",
		"params": map[string]any{"name": "echo", "arguments": map[string]any{"text": "hi"}}})
	result := resp["result"].(map[string]any)
	content := result["content"].([]any)[0].(map[string]any)
	if content["text"] != "hi" {
		t.Fatalf("expected hi, got %v", content["text"])
	}
}

func TestMcpUnknownTool(t *testing.T) {
	bridge := NewMcpBridge(nil)
	resp := bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "id": 4, "method": "tools/call",
		"params": map[string]any{"name": "missing", "arguments": map[string]any{}}})
	errObj := resp["error"].(map[string]any)
	if errObj["code"] != -32602 {
		t.Fatalf("expected -32602, got %v", errObj["code"])
	}
}

func TestMcpUnknownMethod(t *testing.T) {
	bridge := NewMcpBridge(nil)
	resp := bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "id": 5, "method": "bogus"})
	errObj := resp["error"].(map[string]any)
	if errObj["code"] != -32601 {
		t.Fatalf("expected -32601, got %v", errObj["code"])
	}
}

func TestMcpMalformedRequest(t *testing.T) {
	bridge := NewMcpBridge(nil)
	resp := bridge.HandleRequest(map[string]any{})
	errObj := resp["error"].(map[string]any)
	if errObj["code"] != -32600 {
		t.Fatalf("expected -32600, got %v", errObj["code"])
	}
}

type collectSink struct {
	mu     sync.Mutex
	events []map[string]any
}

func newCollectSink() *collectSink {
	return &collectSink{}
}

func (s *collectSink) Send(event map[string]any) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.events = append(s.events, event)
	return nil
}

func (s *collectSink) types() []string {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]string, 0, len(s.events))
	for _, e := range s.events {
		out = append(out, e["type"].(string))
	}
	return out
}

func TestMcpAsyncToolHandlerLifecycle(t *testing.T) {
	sink := newCollectSink()
	bridge := NewMcpBridge(sink)
	bridge.RegisterTool(AsyncToolHandler("build", "build tool", func(args map[string]any, tracker *aep.TaskTracker) map[string]any {
		return map[string]any{"artifact": "app.bin"}
	}))
	resp := bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "id": 1, "method": "tools/call",
		"params": map[string]any{"name": "build", "arguments": map[string]any{"_task_id": "task_x"}}})
	if resp["result"] == nil {
		t.Fatal("expected result")
	}
	if !waitForType(sink, "task.completed", 2*time.Second) {
		t.Fatalf("timed out; got %v", sink.types())
	}
	types := sink.types()
	for _, want := range []string{"task.accepted", "task.started", "task.progress", "task.completed"} {
		if !containsStr(types, want) {
			t.Fatalf("missing %s in %v", want, types)
		}
	}
}

func TestMcpAsyncToolHandlerFailure(t *testing.T) {
	sink := newCollectSink()
	bridge := NewMcpBridge(sink)
	bridge.RegisterTool(AsyncToolHandler("build", "build tool", func(args map[string]any, tracker *aep.TaskTracker) map[string]any {
		panic("build failed")
	}))
	bridge.HandleRequest(map[string]any{"jsonrpc": "2.0", "id": 1, "method": "tools/call",
		"params": map[string]any{"name": "build", "arguments": map[string]any{"_task_id": "task_y"}}})
	if !waitForType(sink, "task.failed", 2*time.Second) {
		t.Fatalf("timed out; got %v", sink.types())
	}
}

func waitForType(sink *collectSink, typ string, timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		if containsStr(sink.types(), typ) {
			return true
		}
		time.Sleep(20 * time.Millisecond)
	}
	return false
}

func containsStr(list []string, s string) bool {
	for _, v := range list {
		if v == s {
			return true
		}
	}
	return false
}

File: implementations/java/src/test/java/com/axisrobo/harmovela/bridge/McpBridgeTest.java
package com.axisrobo.harmovela.bridge;

import org.junit.jupiter.api.Test;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;
import static org.junit.jupiter.api.Assertions.*;

class McpBridgeTest {
    @Test
    @SuppressWarnings("unchecked")
    void initializeReturnsServerInfo() {
        var bridge = new McpBridge(null);
        var resp = bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 1, "method", "initialize", "params", Map.of()));
        var result = (Map<String, Object>) resp.get("result");
        var info = (Map<String, Object>) result.get("serverInfo");
        assertEquals("harmovela-mcp-bridge", info.get("name"));
        assertEquals("0.1.0", result.get("protocolVersion"));
    }

    @Test
    void notificationsInitializedReturnsNull() {
        var bridge = new McpBridge(null);
        assertNull(bridge.handleRequest(Map.of("jsonrpc", "2.0", "method", "notifications/initialized")));
    }

    @Test
    @SuppressWarnings("unchecked")
    void toolsListReturnsRegisteredTools() {
        var bridge = new McpBridge(null);
        bridge.registerTool("echo", Map.of("description", "echo tool", "required", List.of("text")),
            (args, ctx) -> Map.of("content", List.of(Map.of("type", "text", "text", args.get("text")))));
        var resp = bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 2, "method", "tools/list"));
        var result = (Map<String, Object>) resp.get("result");
        var tools = (List<Map<String, Object>>) result.get("tools");
        assertEquals(1, tools.size());
        assertEquals("echo", tools.get(0).get("name"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void toolsCallInvokesHandler() {
        var bridge = new McpBridge(null);
        bridge.registerTool("echo", Map.of("description", "echo"),
            (args, ctx) -> Map.of("content", List.of(Map.of("type", "text", "text", args.get("text")))));
        var resp = bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 3, "method", "tools/call",
            "params", Map.of("name", "echo", "arguments", Map.of("text", "hi"))));
        var result = (Map<String, Object>) resp.get("result");
        var content = (List<Map<String, Object>>) result.get("content");
        assertEquals("hi", content.get(0).get("text"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void unknownToolReturnsError() {
        var bridge = new McpBridge(null);
        var resp = bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 4, "method", "tools/call",
            "params", Map.of("name", "missing", "arguments", Map.of())));
        var error = (Map<String, Object>) resp.get("error");
        assertEquals(-32602, error.get("code"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void unknownMethodReturnsError() {
        var bridge = new McpBridge(null);
        var resp = bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 5, "method", "bogus"));
        var error = (Map<String, Object>) resp.get("error");
        assertEquals(-32601, error.get("code"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void malformedRequestReturnsError() {
        var bridge = new McpBridge(null);
        var resp = bridge.handleRequest(Map.of());
        var error = (Map<String, Object>) resp.get("error");
        assertEquals(-32600, error.get("code"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void handlerExceptionReturnsIsError() {
        var bridge = new McpBridge(null);
        bridge.registerTool("boom", Map.of(), (args, ctx) -> { throw new RuntimeException("kaboom"); });
        var resp = bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 6, "method", "tools/call",
            "params", Map.of("name", "boom", "arguments", Map.of())));
        var result = (Map<String, Object>) resp.get("result");
        assertEquals(true, result.get("isError"));
    }

    @Test
    void asyncToolHandlerEmitsLifecycle() throws Exception {
        var events = new CopyOnWriteArrayList<Map<String, Object>>();
        var bridge = new McpBridge(events::add);
        bridge.registerTool(McpBridge.asyncToolHandler("build", "build tool",
            (args, tracker) -> Map.of("artifact", "app.bin")));
        bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 1, "method", "tools/call",
            "params", Map.of("name", "build", "arguments", Map.of("_task_id", "task_x"))));
        assertTrue(waitForType(events, "task.completed"));
        var types = types(events);
        assertTrue(types.contains("task.accepted"));
        assertTrue(types.contains("task.started"));
        assertTrue(types.contains("task.progress"));
        assertTrue(types.contains("task.completed"));
    }

    @Test
    void asyncToolHandlerEmitsFailedOnError() throws Exception {
        var events = new CopyOnWriteArrayList<Map<String, Object>>();
        var bridge = new McpBridge(events::add);
        bridge.registerTool(McpBridge.asyncToolHandler("build", "build tool",
            (args, tracker) -> { throw new RuntimeException("build failed"); }));
        bridge.handleRequest(Map.of("jsonrpc", "2.0", "id", 1, "method", "tools/call",
            "params", Map.of("name", "build", "arguments", Map.of("_task_id", "task_y"))));
        assertTrue(waitForType(events, "task.failed"));
    }

    private static boolean waitForType(List<Map<String, Object>> events, String type) throws InterruptedException {
        var deadline = System.currentTimeMillis() + 2000;
        while (System.currentTimeMillis() < deadline) {
            if (types(events).contains(type)) return true;
            Thread.sleep(20);
        }
        return false;
    }

    private static List<String> types(List<Map<String, Object>> events) {
        var out = new ArrayList<String>();
        for (var e : events) out.add((String) e.get("type"));
        return out;
    }
}
